### PR TITLE
More comprehensive use of unification when pretyping lambdas and applications

### LINKED ISSUE #11189: Canonical projection for `_ -> _` cannot be used with `fun .. => ..` terms, only with (local) definitions
#### Description of the problem
The following code shows a structure with one type field `key` and canonical structures for a base case (`unit`) and a case for implications `_ -> _`. I had expected the line prefixed with `Fail` below to succeed but it does not. Instead, it reports the following error message:
```
The type of this term is a product while it is expected to be
 (key ?n).
```
I find this very surprising since the code registers a canonical projection for (non-dependent) function types, as Coq gladly reports in `Print Canonical Projections`:
```
_ -> _ <- key ( @step )
```
The reason is not, I think, a failure in CS search but rather a limitation in the pretyper: `evardefine.split_tycon` is what my backtrace points to.

The last four lines show that the problem can be circumvented in multiple ways, including wrapping the `fun` term in `id`.

Is there a reason why it cannot work with `fun`?

```coq
Structure trivial_cs := { key : Type }.

Canonical Structure base : trivial_cs := {|key := unit|}.
Canonical Structure step {A : Type} (t : trivial_cs) : trivial_cs := {| key := A -> key t |}.

Definition test := fun (t : trivial_cs) (k : key t) => tt.

Fail Check test (_) (fun x : unit => tt).

Check let this := (fun x : unit => tt) in test (_) this.
Definition this := fun x : unit => tt.
Check test (_) this.
Check test _ (id (fun x : unit => tt)).
```

#### Coq Version
8.11 branch as of a few days ago.

### PR DESCRIPTION
**Pull request by SkySkimmer**
Replace https://github.com/coq/coq/pull/12383
Fixes #11189

This PR is less careful about perf but let's first have a bench to tell how careful we need to be here.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
https://gitlab.com/coq/coq/-/jobs/1455137062

**Comment by Janno** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-rewriter │  682.54   684.34  -0.26 │  1895342696715   1899786483761  -0.23 │  2931629805833   2929353519595   0.08 │ 1074516  1074404   0.01 │   0    0    nan │
│      coq-mathcomp-odd-order │ 1049.25  1051.16  -0.18 │  2923359415035   2928144133893  -0.16 │  4626016232474   4622387131232   0.08 │  908732   941604  -3.49 │   0    0    nan │
│                  coq-stdlib │  411.72   412.40  -0.16 │  1150915064352   1153182273766  -0.20 │  1466822739834   1466685514088   0.01 │  610320   602544   1.29 │   0    0    nan │
│                    coq-hott │  296.36   296.40  -0.01 │   816777679319    814061542790   0.33 │  1267025942395   1266815464050   0.02 │  570956   570876   0.01 │   0    0    nan │
│        coq-mathcomp-algebra │  140.84   140.81   0.02 │   391305183657    390730216251   0.15 │   510166416533    509419363060   0.15 │  555644   556760  -0.20 │   0    0    nan │
│             coq-fiat-crypto │ 4799.83  4796.16   0.08 │ 13333087726192  13319423662865   0.10 │ 23543338498738  23540298117189   0.01 │ 2223940  2222380   0.07 │   0    0    nan │
│  coq-performance-tests-lite │ 1542.98  1541.78   0.08 │  4307553599551   4301551645626   0.14 │  7116788259597   7116387522348   0.01 │ 2085128  2085112   0.00 │   0    0    nan │
│                   coq-flocq │  146.19   146.07   0.08 │   404984711260    404657009897   0.08 │   503218321131    503232481770  -0.00 │  810152   810076   0.01 │   0    0    nan │
│                    coq-corn │ 1458.34  1456.08   0.16 │  4062898081768   4053706567008   0.23 │  5864694112355   5862667644281   0.03 │  789988   790600  -0.08 │   0    0    nan │
│       coq-mathcomp-solvable │  168.77   168.49   0.17 │   469709284812    468968185740   0.16 │   665752951734    665140806693   0.09 │  691488   692524  -0.15 │   0    0    nan │
│       coq-mathcomp-fingroup │   47.82    47.74   0.17 │   132706732628    132701621171   0.00 │   186307858205    186001706828   0.16 │  483252   480448   0.58 │   0    0    nan │
│                 coq-bignums │   55.65    55.55   0.18 │   153682273877    153343512561   0.22 │   206562973155    206559697735   0.00 │  463604   463580   0.01 │   0    0    nan │
│            coq-fiat-parsers │  581.22   580.06   0.20 │  1626051691056   1620089411478   0.37 │  2490113166737   2487304306144   0.11 │ 3085020  3085088  -0.00 │   0    0    nan │
│                  coq-geocoq │ 1318.41  1315.00   0.26 │  3676004568120   3668534662738   0.20 │  5418705584066   5416953008155   0.03 │ 1037784   979760   5.92 │   0    0    nan │
│       coq-engine-bench-lite │  313.29   312.41   0.28 │   881808385874    879135496030   0.30 │  1643839782481   1646372151969  -0.15 │ 2868224  2868084   0.00 │   0    0    nan │
│              coq-verdi-raft │ 1093.73  1090.54   0.29 │  3046449346784   3036815985401   0.32 │  4306080497782   4299865439516   0.14 │  919416   921052  -0.18 │   0    0    nan │
│          coq-mathcomp-field │  221.36   220.65   0.32 │   615381721244    614872366181   0.08 │   938347629439    938019909739   0.03 │  663180   665000  -0.27 │   0    0    nan │
│      coq-mathcomp-character │  149.54   148.98   0.38 │   415323875331    414237227240   0.26 │   583469386666    583077033759   0.07 │  721228   726240  -0.69 │   0    0    nan │
│                 coq-unimath │ 4319.58  4302.25   0.40 │ 12020760476996  11964924408776   0.47 │ 22900699485962  22902698456002  -0.01 │ 3735232  3740184  -0.13 │   0    0    nan │
│                coq-bedrock2 │  310.16   308.91   0.40 │   859957693062    856840118969   0.36 │  1507106376867   1506196617194   0.06 │  800144   799828   0.04 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │ 1635.29  1626.62   0.53 │  4550514850334   4527344703309   0.51 │  7152014510182   7148186756056   0.05 │ 1074388  1074312   0.01 │   0    0    nan │
│               coq-fourcolor │ 2541.25  2527.56   0.54 │  7073607389855   7045828215866   0.39 │ 12241159179366  12229409703647   0.10 │  732372   732368   0.00 │   0    0    nan │
│            coq-math-classes │  172.22   171.28   0.55 │   481267963612    478268374749   0.63 │   640623402313    640542228945   0.01 │  481960   481960   0.00 │   0    0    nan │
│               coq-perennial │ 6840.11  6802.39   0.55 │ 19054158019170  18950817004942   0.55 │ 29104224431247  29085835231622   0.06 │ 3078448  3067380   0.36 │   0    0    nan │
│                   coq-color │  483.54   480.44   0.65 │  1347754305205   1341792588085   0.44 │  1697224062568   1696020122318   0.07 │ 1154168  1154968  -0.07 │   0    0    nan │
│      coq-mathcomp-ssreflect │   54.03    53.59   0.82 │   147083428634    146586441236   0.34 │   190766001736    190232960608   0.28 │  526956   526052   0.17 │   0    0    nan │
│                 coq-coqutil │   59.45    58.96   0.83 │   164515283709    162959852491   0.95 │   216393094078    215616669431   0.36 │  599976   600076  -0.02 │   0    0    nan │
│              coq-coquelicot │   68.93    68.35   0.85 │   187499186474    186109745140   0.75 │   241044657489    239389564999   0.69 │  557224   557216   0.00 │   0    0    nan │
│                    coq-core │  107.50   106.58   0.86 │   298191655982    297388141100   0.27 │   461063487411    460977340858   0.02 │  330280   330332  -0.02 │   0    0    nan │
│               coq-fiat-core │  108.28   107.12   1.08 │   305529000879    303125261334   0.79 │   412478783869    409686413158   0.68 │  480628   480900  -0.06 │   0    0    nan │
│                coq-coqprime │  331.36   326.68   1.43 │   919956039325    908466983063   1.26 │  1597775844674   1594360461854   0.21 │  918068   918508  -0.05 │   0    0    nan │
│                   coq-verdi │   98.00    95.55   2.56 │   269659962522    262492692394   2.73 │   379173380634    373351180077   1.56 │  542464   542312   0.03 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘
```

Looks worryingly different from the most recent one in #12383.

**Comment by ppedrot** [comment]
@Janno I don't think there is anything worrying here. coqprime is notoriously noisy, and it spends the vast majority of its time in stuff unrelated to unification (namely, the lia solver). Meanwhile, verdi has been experiencing quite a bit of variability recently, and it's not the first time I witness such a time diff. [Last time](https://github.com/coq/coq/pull/14700) it vanished when rerun. So this all looks like noise to me.

**Comment by Janno** [comment]
Ah, I see. Indeed, the problem with coq-verdi seems to be the same as in #14700. I should really abstain from commenting on benchmarks..

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
How does this interact with #14616?

**Comment by SkySkimmer** [comment]
>How does this interact with #14616?

No conflict or easy rebase

**Comment by Janno** [comment]
I think my most recent comment from #12383 applies to this PR as well: https://github.com/coq/coq/pull/12383#issuecomment-889017482. It has two new (failing) test cases that I think should pass. So before we merge this less specific version of #12383 here we should fix those, I think.

File: pretyping/coercion.ml [hunk 1]
```diff
@@ -448,29 +458,41 @@ let mu env sigma t =
...
+  let t = whd_all env sigma j.uj_type in
+  match EConstr.kind sigma t with
+  | Prod _ -> Inl sigma
+  | Evar ev ->
+    let sigma, _ = Evardefine.define_evar_as_product env sigma ev in
+    Inl sigma
+  | _ ->
+    let sigma, (domain_hole, _) = Evarutil.new_type_evar env sigma Evd.univ_flexible in
+    let env' = push_rel Context.(Rel.Declaration.LocalAssum (anonR, domain_hole)) env in
+    let sigma, (codomain_hole, _) = Evarutil.new_type_evar env' sigma Evd.univ_flexible in
```
Comment (SkySkimmer) [inline pretyping/coercion.ml:471]: The bug at https://github.com/coq/coq/pull/12383#issuecomment-889017482 was the lack of this push_rel, so the codomain indeed had an ill-typed evar instance.

Comment (Janno) [reply pretyping/coercion.ml:471]: Ah, that makes sense! Thanks!

**Comment by Janno** [comment]
I don't have any objections to merging this instead of #12383. In fact, the code is much nicer! I also don't have any more failing test cases to report. This change is not quite enough to allow me to derive telescopes with canonical structure but the remaining problems are, I think, related to unification heuristics and I'll open separate feature requests/MRs about that.

**Comment by ppedrot** [comment]
@SkySkimmer if you rebase this and let a green CI run, I'll merge it after.

**Pull request merged**